dateutil/utils.py

```diff
@@ -1,6 +1,24 @@
 # -*- coding: utf-8 -*-
 from __future__ import unicode_literals
 
+from datetime import datetime, time
+
+
+def today(tzinfo=None):
+    """
+    Returns a :py:class:`datetime` representing the current day at midnight
+
+    :param tzinfo:
+        The time zone to attach (also used to determine the current day).
+
+    :return:
+        A :py:class:`datetime.datetime` object representing the current day
+        at midnight.
+    """
+
+    dt = datetime.now(tzinfo)
+    return datetime.combine(dt.date(), time(0, tzinfo=tzinfo))
```

Comment:
Responding to [this comment](https://github.com/dateutil/dateutil/pull/475#discussion_r144706735) by @jbrockmendel:

> +1 for today(tzinfo), but I'm very wary of implementing behavior that differs from same-named stdlib behavior. btw:

> ```python
> >>> tz = pytz.timezone('US/Pacific')
> >>> pd.Timestamp.today(tz)
> Timestamp('2017-10-14 17:19:04.081315-0700', tz='US/Pacific')
> ```

My reason for calling it `today` was because, frankly, I think that this is close to `datetime.datetime.today()` *should* do, and I think it violates principle of least surprise to find that `datetime.today()` is basically just an alias for `datetime.now()`. That said, the name overloading *is* my main concern about this function.

I think there are a few reasonable ways forward:

1. Leave as is because this is what `datetime.today()` *should have been* and people are more likely to be shocked that `datetime.datetime.today()` does something else than that `dateutil.utils.today()` does what it does.
2. Change the signature of this so that `utils.today()` takes an optional `time` argument in addition to an optional `tzinfo` argument. This makes it more clear that what it does is give you "today at this time, defaulting to midnight". The reason I don't like this is that there could be a conflict between `time` (which may have a `tzinfo` of its own) and `tzinfo`, and also the interface isn't nearly as clean.
3. Change the name to something like `today_midnight()` or `today_datetime()` or something.

I think I'm inclined to leave it as is, considering `pendulum` has a similar real distinction between its `now` and `today` behavior:

```python
>>> import pendulum
>>> print(pendulum.now())
2017-10-15T17:30:33.887480-04:00
>>>> print(pendulum.today())
2017-10-15T00:00:00-04:00
```